chain/chain/src/flat_storage_resharder.rs

```diff
@@ -0,0 +1,943 @@
+use near_primitives::trie_key::trie_key_parsers::{
+    parse_account_id_from_access_key_key, parse_account_id_from_account_key,
+    parse_account_id_from_contract_code_key, parse_account_id_from_contract_data_key,
+    parse_account_id_from_received_data_key, parse_account_id_from_trie_key_with_separator,
+};
+use near_primitives::types::AccountId;
+use near_store::adapter::flat_store::{FlatStoreAdapter, FlatStoreUpdateAdapter};
+use near_store::adapter::StoreAdapter;
+use near_store::flat::{
+    BlockInfo, FlatStorageReadyStatus, FlatStorageReshardingStatus, FlatStorageStatus,
+    SplittingParentStatus,
+};
+use near_store::{ShardUId, StorageError};
+
+/// `FlatStorageResharder` takes care of updating flat storage when a resharding event
+/// happens.
+///
+/// On an high level, the events supported are:
+/// - #### Shard splitting
+///     Parent shard must be split into two children. The entire operation freezes the flat storage
+///     for the involved shards.
+///     Children shards are created empty and the key-values of the parent will be copied into one of them,
+///     in the background.
+///
+///     After the copy is finished the children shard will have the correct state at some past block height.
+///     It'll be necessary to perform catchup before the flat storage can be put again in Ready state.
+///     The parent shard storage is not needed anymore and can be removed.
+///
+/// The resharder has also the following properties:
+/// - Background processing: the bulk of resharding is done in a separate task, see [FlatStorageResharderScheduler]
+/// - Interruptible: a reshard operation can be interrupted through a [FlatStorageResharderController].
+///     - In the case of event `Split` the state of flat storage will go back to what it was previously.
+#[derive(Clone)]
+pub struct FlatStorageResharder {
+    runtime: Arc<dyn RuntimeAdapter>,
+    resharding_event: Arc<Mutex<Option<FlatStorageReshardingEventStatus>>>,
+}
+
+impl FlatStorageResharder {
+    /// Creates a new `FlatStorageResharder`.
+    pub fn new(runtime: Arc<dyn RuntimeAdapter>) -> Self {
+        let resharding_event = Arc::new(Mutex::new(None));
+        Self { runtime, resharding_event }
+    }
+
+    /// Resumes a resharding event that was in progress.
+    ///
+    /// # Args:
+    /// * `shard_uid`: UId of the shard
+    /// * `status`: resharding status of the shard
+    /// * `scheduler`: component used to schedule the background tasks
+    /// * `controller`: manages the execution of the background tasks
+    pub fn resume(
+        &self,
+        shard_uid: ShardUId,
+        status: &FlatStorageReshardingStatus,
+        scheduler: &dyn FlatStorageResharderScheduler,
+        controller: FlatStorageResharderController,
+    ) -> Result<(), Error> {
+        match status {
+            FlatStorageReshardingStatus::CreatingChild => {
+                // Nothing to do here because the parent will take care of resuming work.
```

Comment:
I wouldn't be surprised to see some cleanup logic here but it's also fine to do it from parent.